Compilation/linking issue

[sorry for filling the issue space:]
still teh same machine as #3648. Thanks to #3649 compilation works, but linking gave an error

       LINK    gap
    obj/src/.libs/collectors.o:(.data.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'

Searching found a recommendation:

If I change `GNUmakefile` in the line

    LIBS = -lm -ldl -lutil -lstdc++

(added the `-lstdc++` bit) then compilation succeeds. Alas I don't understand whether:

a) I can do so without harm for other systems

b) How to do it actually (as the line is auto-created). 

c) Is it just an artifect of me compiling a new autoconf and not compiling a new gcc?

Thank you, @fingolfin for looking at it.
The output is

    Unix54 /home/fac/f/hulpke/gap4 > make V=1
    /bin/sh ./libtool --mode=link gcc -std=gnu99 -export-dynamic  /home/fac/f/hulpke/gap4/extern/install/gmp/lib/libgmp.la       obj/src/ariths.lo obj/src/bags.lo obj/src/blister.lo obj/src/bool.lo obj/src/c_oper1.lo obj/src/c_type1.lo obj/src/calls.lo obj/src/code.lo obj/src/collectors.lo obj/src/compiler.lo obj/src/compstat.lo obj/src/costab.lo obj/src/cyclotom.lo obj/src/debug.lo obj/src/dt.lo obj/src/dteval.lo obj/src/error.lo obj/src/exprs.lo obj/src/ffdata.lo obj/src/finfield.lo obj/src/funcs.lo obj/src/gap.lo obj/gen/gap_version.lo obj/src/gvars.lo obj/src/hookintrprtr.lo obj/src/info.lo obj/src/integer.lo obj/src/intfuncs.lo obj/src/intrprtr.lo obj/src/io.lo obj/src/iostream.lo obj/src/libgap-api.lo obj/src/listfunc.lo obj/src/listoper.lo obj/src/lists.lo obj/src/macfloat.lo obj/src/modules.lo obj/src/modules_builtin.lo obj/src/objcftl.lo obj/src/objects.lo obj/src/objfgelm.lo obj/src/objpcgel.lo obj/src/objset.lo obj/src/opers.lo obj/src/permutat.lo obj/src/plist.lo obj/src/pperm.lo obj/src/precord.lo obj/src/profile.lo obj/src/range.lo obj/src/rational.lo obj/src/read.lo obj/src/records.lo obj/src/saveload.lo obj/src/scanner.lo obj/src/sctable.lo obj/src/set.lo obj/src/stats.lo obj/src/streams.lo obj/src/stringobj.lo obj/src/syntaxtree.lo obj/src/sysfiles.lo obj/src/sysmem.lo obj/src/system.lo obj/src/tietze.lo obj/src/trans.lo obj/src/vars.lo obj/src/vec8bit.lo obj/src/vecffe.lo obj/src/vecgf2.lo obj/src/vector.lo obj/src/weakptr.lo obj/src/gasman.lo   -lz -lreadline    -lm -ldl -lutil -o gap
    libtool: link: gcc -std=gnu99 obj/src/.libs/ariths.o obj/src/.libs/bags.o obj/src/.libs/blister.o obj/src/.libs/bool.o obj/src/.libs/c_oper1.o obj/src/.libs/c_type1.o obj/src/.libs/calls.o obj/src/.libs/code.o obj/src/.libs/collectors.o obj/src/.libs/compiler.o obj/src/.libs/compstat.o obj/src/.libs/costab.o obj/src/.libs/cyclotom.o obj/src/.libs/debug.o obj/src/.libs/dt.o obj/src/.libs/dteval.o obj/src/.libs/error.o obj/src/.libs/exprs.o obj/src/.libs/ffdata.o obj/src/.libs/finfield.o obj/src/.libs/funcs.o obj/src/.libs/gap.o obj/gen/.libs/gap_version.o obj/src/.libs/gvars.o obj/src/.libs/hookintrprtr.o obj/src/.libs/info.o obj/src/.libs/integer.o obj/src/.libs/intfuncs.o obj/src/.libs/intrprtr.o obj/src/.libs/io.o obj/src/.libs/iostream.o obj/src/.libs/libgap-api.o obj/src/.libs/listfunc.o obj/src/.libs/listoper.o obj/src/.libs/lists.o obj/src/.libs/macfloat.o obj/src/.libs/modules.o obj/src/.libs/modules_builtin.o obj/src/.libs/objcftl.o obj/src/.libs/objects.o obj/src/.libs/objfgelm.o obj/src/.libs/objpcgel.o obj/src/.libs/objset.o obj/src/.libs/opers.o obj/src/.libs/permutat.o obj/src/.libs/plist.o obj/src/.libs/pperm.o obj/src/.libs/precord.o obj/src/.libs/profile.o obj/src/.libs/range.o obj/src/.libs/rational.o obj/src/.libs/read.o obj/src/.libs/records.o obj/src/.libs/saveload.o obj/src/.libs/scanner.o obj/src/.libs/sctable.o obj/src/.libs/set.o obj/src/.libs/stats.o obj/src/.libs/streams.o obj/src/.libs/stringobj.o obj/src/.libs/syntaxtree.o obj/src/.libs/sysfiles.o obj/src/.libs/sysmem.o obj/src/.libs/system.o obj/src/.libs/tietze.o obj/src/.libs/trans.o obj/src/.libs/vars.o obj/src/.libs/vec8bit.o obj/src/.libs/vecffe.o obj/src/.libs/vecgf2.o obj/src/.libs/vector.o obj/src/.libs/weakptr.o obj/src/.libs/gasman.o -o gap -Wl,--export-dynamic  /home/fac/f/hulpke/gap4/extern/install/gmp/lib/libgmp.so -lz -lreadline -lm -ldl -lutil -Wl,-rpath -Wl,/home/fac/f/hulpke/gap4/extern/install/gmp/lib -Wl,-rpath -Wl,/home/fac/f/hulpke/gap4/extern/install/gmp/lib
    obj/src/.libs/collectors.o:(.data.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'
    collect2: ld returned 1 exit status
    make: *** [gap] Error 1

Also, just in case it helps:

    Unix55 /home/fac/f/hulpke/gap4 > make -v
    GNU Make 3.81
    Copyright (C) 2006  Free Software Foundation, Inc.
    This is free software; see the source for copying conditions.
    There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
    PARTICULAR PURPOSE.

    This program built for x86_64-redhat-linux-gnu
    Unix56 /home/fac/f/hulpke/gap4 > gcc -v
    Using built-in specs.
    Target: x86_64-redhat-linux
    Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-languages=c,c++,objc,obj-c++,java,fortran,ada --enable-java-awt=gtk --disable-dssi --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-1.5.0.0/jre --enable-libgcj-multifile --enable-java-maintainer-mode --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libjava-multilib --with-ppl --with-cloog --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
    Thread model: posix
    gcc version 4.4.7 20120313 (Red Hat 4.4.7-23) (GCC)
    Unix57 /home/fac/f/hulpke/gap4 > uname -a
    Linux tx1.math.colostate.edu 2.6.32-754.18.2.el6.x86_64 #1 SMP Wed Aug 14 16:26:59 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -18,6 +18,7 @@ environment:
       CYG_ROOT: C:\cygwin64
       PACKAGES: "-P libgmp-devel,zlib-devel,python27,python27-pip"
       CFLAGS: "--coverage"
+      CXXFLAGS: "--coverage"
       LDFLAGS: "--coverage"
 # FIXME: HPC-GAP build on AppVeyor disabled for now, as it crashes
 #    - CYG_ARCH: x86_64

--- a/.travis.yml
+++ b/.travis.yml
@@ -140,6 +140,7 @@ matrix:
 
     # test Julia integration
     - env: TEST_SUITES="testinstall" JULIA=yes CONFIGFLAGS="--disable-Werror"
+      dist: bionic # need GCC 6 to avoid linker error
 
 script:
   - set -e

--- a/Makefile.rules
+++ b/Makefile.rules
@@ -358,7 +358,7 @@ obj/%.lo: %.c cnf/GAP-CFLAGS cnf/GAP-CPPFLAGS libtool
 # Linker rules for gap executable
 ########################################################################
 
-LINK=$(LIBTOOL) --mode=link $(CC) -export-dynamic
+LINK=$(LIBTOOL) --mode=link $(CXX) -export-dynamic
 
 libgap.la: $(OBJS) cnf/GAP-LDFLAGS cnf/GAP-LIBS cnf/GAP-OBJS
 	$(QUIET_LINK)$(LINK) -no-undefined -rpath $(libdir) $(GAP_LDFLAGS) $(OBJS) $(GAP_LIBS) -o $@
@@ -391,8 +391,8 @@ GAP_LDFLAGS += -Wl,--allow-multiple-definition
 all: bin/$(GAPARCH)/gap.dll
 bin/$(GAPARCH)/gap.dll: symlinks libgap.la
 	cp .libs/cyggap-0.dll $@  # FIXME: HACK to support kernel extensions
-gap$(EXEEXT): libgap.la cnf/GAP-LDFLAGS cnf/GAP-LIBS cnf/GAP-OBJS
-	$(QUIET_LINK)$(LINK) $(GAP_LDFLAGS) $(srcdir)/src/gapw95.c $(GAP_LIBS) libgap.la -o $@
+gap$(EXEEXT): obj/src/gapw95.lo libgap.la cnf/GAP-LDFLAGS cnf/GAP-LIBS cnf/GAP-OBJS
+	$(QUIET_LINK)$(LINK) $(GAP_LDFLAGS) obj/src/gapw95.lo $(GAP_LIBS) libgap.la -o $@
 	@( if which peflags > /dev/null ; then peflags --cygwin-heap=2048 gap$(EXEEXT) ; fi )
 
 else
@@ -694,6 +694,9 @@ endif # BUILD_BOEHM_GC
 # ensure subprojects are built and "installed" before compiling and linking GAP
 gap$(EXEEXT): $(EXTERN_FILES)
 $(OBJS): $(EXTERN_FILES)
+ifeq ($(SYS_IS_CYGWIN32),yes)
+obj/src/gapw95.lo: $(EXTERN_FILES)
+endif
 ifeq ($(HPCGAP),yes)
 $(SOURCES): $(EXTERN_FILES)
 endif

